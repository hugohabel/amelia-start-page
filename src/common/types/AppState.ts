import { CountdownFormat } from './amelia';

// App State
export interface IAppState {
  widgets: any[];
  sidebar: ISidebar;
}

// Sidebar
export interface ISidebar {
  isOpen: boolean;
  activeMenu: TActiveMenu;
  activeView: TActiveView;
}

// Countdown Widget
export interface ICountdownWidget {
  type: 'countdown';
  enabled: boolean;
  data: {
    events: IEvent[];
  };
}

export interface IEvent {
  id: string;
  name: string;
  date: string;
  format: CountdownFormat;
}
// End Countdown Widget

// Weather Widget
export interface IWeatherWidget {
  type: 'weather';
  enabled: boolean;
  data: {
    locations: Location[];
  };
}

export interface ILocation {
  id: string;
  lat: string;
  lng: string;
}
// End Weather Widget

// Auxiliary Types
export type TActiveMenu = 'add' | 'about' | 'config' | null;
export type TActiveView = 'add_countdown' | 'add_weather' | null;

// App State Actions
type TAction =
  | { type: 'addNewWidget'; widget: ICountdownWidget | IWeatherWidget }
  | { type: 'toggleSidebar' }
  | { type: 'setActiveMenu'; activeMenu: TActiveMenu }
  | { type: 'setActiveView'; activeView: TActiveView }
  | { type: 'addNewCountdownEvent'; event: IEvent }
  | { type: 'addNewWeatherLocation'; location: Location }
  | { type: 'removeWidgetItem'; uuid: string; widgetType: string };

export type TAppAction = TAction;
export type TAppDispatch = (action: TAppAction) => void;
