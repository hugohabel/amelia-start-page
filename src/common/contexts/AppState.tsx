// External Dependencies
import { createContext, ReactNode, useContext, useReducer } from 'react';
import { debounce } from 'lodash';

// Internal Dependencies
import { LOCAL_STORAGE_STATE_KEY } from '../constants';
import { TAppAction, TAppDispatch, IAppState } from '../types/AppState';
import { SafeLocalStorage } from '../utils/localStorage/localStorage';
import { initialAppState, rootReducer } from '../state/state';

// Interfaces + Types
interface IAppStateProviderProps {
  children: ReactNode;
  initialStateOverride?: IAppState;
}

// Utility Functions

/**
 * Retrieves state from local storage or initializes with default state.
 *
 * @param defaultState AppState   Default app state.
 * @returns AppState  State (can be an existing one, or a default one)
 */
function getStateFromLocalStorageOrInitialize(defaultState: IAppState) {
  const stateFromLocalStorage = SafeLocalStorage.getInstance().getItem(LOCAL_STORAGE_STATE_KEY);
  if (stateFromLocalStorage) {
    return JSON.parse(stateFromLocalStorage) as IAppState;
  }
  return defaultState;
}

/**
 * Persists app state in memory or in local storage
 *
 * @param state AppState  State to be persisted.
 */
function persistState(state: IAppState) {
  SafeLocalStorage.getInstance().setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: false,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: IAppState, action: TAppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

// Create Contexts
const AppStateContext = createContext<IAppState>(initialAppState);
const AppDispatchContext = createContext<TAppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn('Calling default dispatch is a no-op!')
);

/**
 * App State Provider
 *
 * @param param AppStateProviderProps Props to initialize the app state provider.
 * @returns JSX App Dispatch and App State Contexts.
 */
export function AppStateProvider({ children, initialStateOverride }: IAppStateProviderProps) {
  const hydratedInitialAppState = {
    ...initialAppState,
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    getStateFromLocalStorageOrInitialize
  );

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
}

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);
