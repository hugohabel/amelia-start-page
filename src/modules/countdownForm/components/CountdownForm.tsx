// External Dependencies
import { useState } from 'react';
import { FormControlLabel, Radio, RadioGroup } from '@material-ui/core';
import * as Yup from 'yup';
import { useFormik } from 'formik';
import { v4 as uuidv4 } from 'uuid';
import dayjs from 'dayjs';
import EmojiPicker, { IEmojiData } from 'emoji-picker-react';

// Internal Dependencies
import { CountdownFormat } from '../../../common/types/amelia';
import { useAppDispatch } from '../../../common/contexts/AppState';

// Styles
import styles from './CountdownForm.module.css';
import sharedStyles from '../../../common/styles/Shared.module.css';

/**
 * Countdown Form
 *
 * Form used to create a new countdown.
 *
 * @returns JSX
 */
function CountdownForm() {
  // Component's State
  const [chosenEmoji, setChosenEmoji] = useState<IEmojiData>();
  const [displayEmojiPicker, setDisplayEmojiPicker] = useState(false);

  // Hooks
  const dispatch = useAppDispatch();

  // Handlers

  /* istanbul ignore next */
  const onEmojiClickHandler = (event: any, emojiObject: IEmojiData) => {
    setChosenEmoji(emojiObject);
    setDisplayEmojiPicker(false);
  };

  const onEmojiPlaceholderClick = () => {
    setDisplayEmojiPicker(true);
  };

  // Form rules and validations (Managed with Formik)
  const formik = useFormik({
    initialValues: {
      eventName: '',
      eventDate: '',
      eventFormat: 'days+hours',
    },
    validationSchema: Yup.object({
      eventName: Yup.string()
        .min(2, 'Must be 2 characters or more')
        .required('Event name is required'),
      eventDate: Yup.string().required('Event date is required'),
    }),
    onSubmit: (values) => {
      dispatch({
        type: 'addNewCountdownEvent',
        event: {
          id: uuidv4(),
          name: values.eventName,
          date: dayjs(values.eventDate).toISOString(),
          format: values.eventFormat as CountdownFormat,
          emoji: chosenEmoji?.emoji || 'ðŸ’¡',
        },
      });
      dispatch({
        type: 'setActiveView',
        activeView: null,
      });
    },
  });

  return (
    <form onSubmit={formik.handleSubmit} autoComplete="off">
      <div className={styles.inputContainer}>
        <div
          data-testid="event-emoji"
          onClick={onEmojiPlaceholderClick}
          className={styles.emojiPlaceholder}>
          {chosenEmoji ? chosenEmoji?.emoji : 'ðŸ’¡'}
        </div>
        {/* Event Name + Error Display */}
        <input
          type="text"
          id="eventName"
          name="eventName"
          placeholder="Event name"
          className={styles.eventNameInput}
          onChange={formik.handleChange}
          value={formik.values.eventName}
        />
        {/* End Event Name + Error Display */}
      </div>
      {formik.touched.eventName && formik.errors.eventName ? (
        <div className={`${styles.formError} ${sharedStyles.errorText}`}>
          {formik.errors.eventName}
        </div>
      ) : null}
      {displayEmojiPicker && (
        <EmojiPicker
          pickerStyle={{
            'box-shadow': 'none',
            width: '328px',
            'font-family': 'PT Sans',
            'margin-bottom': '16px',
            background: '#414558',
            color: '#FFFFFF',
            border: '1px solid transparent',
          }}
          onEmojiClick={onEmojiClickHandler}
          disableSearchBar
        />
      )}

      {/* Event Date + Error Display */}
      <input
        type="date"
        id="eventDate"
        name="eventDate"
        placeholder="Event date"
        onChange={formik.handleChange}
        value={formik.values.eventDate}
      />
      {formik.touched.eventDate && formik.errors.eventDate ? (
        <div className={`${styles.formError} ${sharedStyles.errorText}`}>
          {formik.errors.eventDate}
        </div>
      ) : null}
      {/* End Event Date + Error Display */}

      {/* Event Format */}
      <h4 className={styles.subtitle}>How do you want to display the countdown?</h4>
      <RadioGroup
        defaultValue={formik.initialValues.eventFormat}
        value={formik.values.eventFormat.toString()}
        name="eventFormat"
        className={styles.radioGroup}
        onChange={(event) => {
          formik.setFieldValue('eventFormat', event.currentTarget.value);
        }}>
        <FormControlLabel
          value="days+hours"
          control={<Radio style={{ color: '#9580FF' }} />}
          label="Days and Hours"
        />
        <FormControlLabel
          value="days"
          control={<Radio style={{ color: '#9580FF' }} />}
          label="Days"
        />
      </RadioGroup>
      {/* End Event Format */}

      <button type="submit">Submit</button>
    </form>
  );
}

export { CountdownForm };
