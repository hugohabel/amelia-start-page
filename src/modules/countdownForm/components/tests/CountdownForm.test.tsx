// External Dependencies
import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

// Internal Dependencies
import { AppStateProvider, useAppDispatch } from '../../../../common/contexts/AppState';
import CountdownForm from '../CountdownForm';

// Mocked Functions
jest.mock('../../../../common/contexts/AppState', () => ({
  ...(jest.requireActual('../../../../common/contexts/AppState') as object),
  useAppDispatch: jest.fn(),
}));
let mockAppDispatch: ReturnType<typeof useAppDispatch>;

// Mocked Values
const EVENTNAMEVALUE = 'Christmas';

describe('Countdown Form', () => {
  it('matches snapshot', async () => {
    const component = render(
      <CountdownForm />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  describe('Countdown Form - Handlers', () => {

    beforeEach(() => {
      mockAppDispatch = jest.fn();
      (useAppDispatch as jest.Mock).mockReturnValue(mockAppDispatch);
    });

    it('triggers onSubmit when clicking on the submit button', async () => {
      const component = render(
        <AppStateProvider>
          <CountdownForm />
        </AppStateProvider>
      );

      // Get Submit Button
      const submitButtonElement = await component.findByRole('button', { name: 'Submit' });
      expect(submitButtonElement).toBeInTheDocument();

      // Get Event Name Input
      const eventNameInputElement = await component.findByPlaceholderText('Event name');
      expect(eventNameInputElement).toBeInTheDocument();

      // Set Form Values
      fireEvent.change(eventNameInputElement, { target: { value: EVENTNAMEVALUE } });

      // Trigger Click
      fireEvent.click(submitButtonElement);

      await waitFor(() => {
        expect(mockAppDispatch).toHaveBeenCalledTimes(2);
      });
    });

    it('displays validation error when clicking on the submit button with empty event name', async () => {
      const component = render(
        <AppStateProvider>
          <CountdownForm />
        </AppStateProvider>
      );

      // Get Submit Button
      const submitButtonElement = await component.findByRole('button', { name: 'Submit' });
      expect(submitButtonElement).toBeInTheDocument();

      // Trigger Click
      fireEvent.click(submitButtonElement);

      // Get Event Name Error
      const eventNameErrorElement = await component.findByText('Event name is required');
      expect(eventNameErrorElement).toBeInTheDocument();
    });
  });
});