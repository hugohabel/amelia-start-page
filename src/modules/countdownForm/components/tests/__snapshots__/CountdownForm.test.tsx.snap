// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown Form matches snapshot 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
  >
    <div
      class="inputContainer"
    >
      <div
        class="emojiPlaceholder"
      >
        ðŸ’¡
      </div>
      <input
        class="eventNameInput"
        id="eventName"
        name="eventName"
        placeholder="Event name"
        type="text"
        value=""
      />
    </div>
    <input
      id="eventDate"
      name="eventDate"
      placeholder="Event date"
      type="date"
      value=""
    />
    <h4
      class="subtitle"
    >
      How do you want to display the countdown?
    </h4>
    <div
      class="MuiFormGroup-root radioGroup"
      role="radiogroup"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorSecondary"
          style="color: rgb(149, 128, 255);"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input-4"
              name="eventFormat"
              type="radio"
              value="days+hours"
            />
            <div
              class="PrivateRadioButtonIcon-root-5 PrivateRadioButtonIcon-checked-7"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Days and Hours
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
          style="color: rgb(149, 128, 255);"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-4"
              name="eventFormat"
              type="radio"
              value="days"
            />
            <div
              class="PrivateRadioButtonIcon-root-5"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Days
        </span>
      </label>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
